src/Chat/ChatEditor.ts

```diff
@@ -0,0 +1,83 @@
+import { Chat, ChatMessage, ChatRole, MessageBlock } from "../../shared/types";
+import { stringToMessageBlocks } from "../../shared/utils/messageBlockHelpers";
+import { ExtensionToWebviewMessageSender } from "../webview/ExtensionToWebviewMessageSender";
+import { ChatDataManager } from "./ChatDataManager";
+
+/**
+ * Handles editing chat content, performing corresponding updates to the ChatDataManager and ChatWebviewProvider.
+ */
+export class ChatEditor {
+  private readonly chatDataManager: ChatDataManager;
+  private readonly extensionToWebviewMessageSender: ExtensionToWebviewMessageSender;
+
+  public constructor(
+    chatDataManager: ChatDataManager,
+    extensionToWebviewMessageSender: ExtensionToWebviewMessageSender,
+  ) {
+    this.chatDataManager = chatDataManager;
+    this.extensionToWebviewMessageSender = extensionToWebviewMessageSender;
+  }
+
+  /**
+   * Appends MessageBlocks to the given chat with the given role.
+   * Adds to the last message if it has the same role or creates a new message.
+   * Performs the necessary updates to the chat in the ChatDataManager.
+   * If updateWebview is true, the webview will be updated after appending the message blocks.
+   * @param chat Chat to update.
+   * @param role ChatRole of the given content.
+   * @param messageBlocks MessageBlocks to append to the chat.
+   * @param updateWebview Whether to update the webview after appending the message blocks.
+   */
+  public async appendMessageBlocks(
+    chat: Chat,
+    role: ChatRole,
+    messageBlocks: MessageBlock[],
+    updateWebview: boolean = true,
+  ) {
+    const lastMessage: ChatMessage = chat.messages[chat.messages.length - 1];
+    if (lastMessage && lastMessage.role === role) {
+      lastMessage.content = lastMessage.content.concat(messageBlocks);
+    } else {
```

Comment:
Also, if we keep it, I've add that reasoning in comments

Reply:
To me it makes sense to extend the prior message if the roles are the same, especially since some LLMs don't allow for consecutive messages with the same role.